x/staking/handler_test.go

```diff
@@ -0,0 +1,307 @@
+package staking_test
+
+import (
+	//"errors"
+	"testing"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	. "github.com/smartystreets/goconvey/convey"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	"github.com/KuChainNetwork/kuchain/chain/config"
+	"github.com/KuChainNetwork/kuchain/chain/constants"
+	"github.com/KuChainNetwork/kuchain/chain/types"
+	"github.com/KuChainNetwork/kuchain/test/simapp"
+	stakingTypes "github.com/KuChainNetwork/kuchain/x/staking/types"
+	"github.com/tendermint/tendermint/crypto"
+)
+
+func NewTestApp(wallet *simapp.Wallet) (addAlice, addJack, addValidator sdk.AccAddress, accAlice, accJack, accValidator types.AccountID, app *simapp.SimApp) {
+	addAlice = wallet.NewAccAddress()
+	addJack = wallet.NewAccAddress()
+	addValidator = wallet.NewAccAddress()
+
+	accAlice = types.MustAccountID("alice@ok")
+	accJack = types.MustAccountID("jack@ok")
+	accValidator = types.MustAccountID("validator@ok")
+
+	resInt, succ := sdk.NewIntFromString("100000000000000000000000")
+	if !succ {
+		resInt = sdk.NewInt(10000000000000000)
+	}
+	initAsset := types.NewCoin(constants.DefaultBondDenom, resInt)
+	asset1 := types.Coins{
+		types.NewInt64Coin("foo/coin", 67),
+		initAsset}
+
+	asset2 := types.Coins{
+		types.NewInt64Coin("foo/coin", 67),
+		types.NewInt64Coin(constants.DefaultBondDenom, 10000000)}
+
+	genAlice := simapp.NewSimGenesisAccount(accAlice, addAlice).WithAsset(asset1)
+	genJack := simapp.NewSimGenesisAccount(accJack, addJack).WithAsset(asset1)
+	genValidator := simapp.NewSimGenesisAccount(accValidator, addValidator).WithAsset(asset2)
+
+	genAccs := simapp.NewGenesisAccounts(wallet.GetRootAuth(), genAlice, genJack, genValidator)
+	app = simapp.SetupWithGenesisAccounts(genAccs)
+
+	ctxCheck := app.BaseApp.NewContext(true, abci.Header{Height: app.LastBlockHeight() + 1})
+
+	accountAlice := app.AccountKeeper().GetAccount(ctxCheck, accAlice)
+	accountJack := app.AccountKeeper().GetAccount(ctxCheck, accJack)
+	accountValidator := app.AccountKeeper().GetAccount(ctxCheck, accValidator)
+
+	So(accountAlice, ShouldNotBeNil)
+	So(genAlice.GetID().Eq(accountAlice.GetID()), ShouldBeTrue)
+	So(genAlice.GetAuth().Equals(accountAlice.GetAuth()), ShouldBeTrue)
+
+	So(accountJack, ShouldNotBeNil)
+	So(genJack.GetID().Eq(accountJack.GetID()), ShouldBeTrue)
+	So(genJack.GetAuth().Equals(accountJack.GetAuth()), ShouldBeTrue)
+
+	So(accountValidator, ShouldNotBeNil)
+	So(genValidator.GetID().Eq(accountValidator.GetID()), ShouldBeTrue)
+	So(genValidator.GetAuth().Equals(accountValidator.GetAuth()), ShouldBeTrue)
+
+	return addAlice, addJack, addValidator, accAlice, accJack, accValidator, app
+}
+
+func CreateValidator(t *testing.T, wallet *simapp.Wallet, app *simapp.SimApp, addAlice sdk.AccAddress, accAlice types.AccountID, rate sdk.Dec, pk crypto.PubKey, passed bool) error {
```

Comment:
Should be createValudator if no use by others

Reply:
Modified